src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -9,6 +9,63 @@ namespace System.Collections.Generic
     /// <summary>Internal helper functions for working with enumerables.</summary>
     internal static class EnumerableHelpers
     {
+        /// <summary>
+        /// Copies items from an enumerable to an array.
+        /// </summary>
+        /// <typeparam name="T">The element type of the enumerable.</typeparam>
+        /// <param name="source">The source enumerable.</param>
+        /// <param name="array">The destination array.</param>
+        /// <param name="arrayIndex">The index in the array to start copying to.</param>
+        /// <param name="count">The number of items in the enumerable.</param>
+        internal static void Copy<T>(IEnumerable<T> source, T[] array, int arrayIndex, int count)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(arrayIndex >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert(array?.Length - arrayIndex >= count);
+
+            var collection = source as ICollection<T>;
+            if (collection != null)
+            {
+                Debug.Assert(collection.Count == count);
+                collection.CopyTo(array, arrayIndex);
+                return;
+            }
+
+            IterativeCopy(source, array, arrayIndex, count);
+        }
+
+        /// <summary>
+        /// Copies items from a non-collection enumerable to an array.
+        /// </summary>
+        /// <typeparam name="T">The element type of the enumerable.</typeparam>
+        /// <param name="source">The source enumerable.</param>
+        /// <param name="array">The destination array.</param>
+        /// <param name="arrayIndex">The index in the array to start copying to.</param>
+        /// <param name="count">The number of items in the enumerable.</param>
+        internal static void IterativeCopy<T>(IEnumerable<T> source, T[] array, int arrayIndex, int count)
+        {
+            Debug.Assert(source != null && !(source is ICollection<T>));
+            Debug.Assert(arrayIndex >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert(array?.Length - arrayIndex >= count);
+
+            using (IEnumerator<T> enumerator = source.GetEnumerator())
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    if (!enumerator.MoveNext())
+                    {
+                        Debug.Fail($"Expected length of {nameof(source)} to be exactly {count}.");
+                    }
+
+                    array[arrayIndex + i] = enumerator.Current;
+                }
+
+                Debug.Assert(!enumerator.MoveNext(), $"Expected length of {nameof(source)} to be exactly {count}.");
```

Comment:
@JonHanna While it is true that `count` is only of debug value, I don't think think this function semantically makes sense without it. If you're saying it's safe to copy this enumerable to this array, then you would have to know the count of the enumerable (or at least an upper bound thereof) in advance. I doubt there will ever be a callsite for this function that could not provide a count.